File: 01-Languages/Go/11-OOP/main.go
// main.go - structs+methods, implicit interface satisfaction, struct embedding.
package main

import "fmt"

type Animal struct {
	Name string
}

func (a Animal) Speak() string {
	return a.Name + " makes a sound"
}

type Speaker interface {
	Speak() string
}

func Announce(s Speaker) {
	fmt.Println(s.Speak())
}

type Dog struct {
	Animal
	Breed string
}

func main() {
	fmt.Println("--- struct with a method ---")
	rex := Animal{Name: "Rex"}
	fmt.Println(rex.Speak())

	fmt.Println("\n--- implicit interface satisfaction (no 'implements' keyword) ---")
	Announce(rex)

	fmt.Println("\n--- struct embedding: composition, not inheritance ---")
	d := Dog{Animal: Animal{Name: "Fido"}, Breed: "Labrador"}
	fmt.Println("d.Speak() (promoted from Animal):", d.Speak())
	fmt.Println("d.Name (promoted field):", d.Name)
	fmt.Println("d.Breed (Dog's own field):", d.Breed)
	Announce(d) // Dog ALSO satisfies Speaker, via the promoted method
}

File: 01-Languages/Rust/11-OOP/main.rs
// main.rs - structs+impl, explicit trait implementation, trait default methods, composition.

struct Animal {
    name: String,
}

impl Animal {
    fn new(name: &str) -> Self {
        Animal { name: name.to_string() }
    }
}

trait Speaker {
    fn speak(&self) -> String;
}

impl Speaker for Animal {
    fn speak(&self) -> String {
        format!("{} makes a sound", self.name)
    }
}

fn announce(s: &impl Speaker) {
    println!("{}", s.speak());
}

trait Greet {
    fn name(&self) -> String;
    fn greeting(&self) -> String {
        format!("Hello, {}!", self.name())
    }
}

impl Greet for Animal {
    fn name(&self) -> String {
        self.name.clone()
    }
    // greeting() is NOT overridden -- uses the trait's default implementation
}

struct Dog {
    animal: Animal,
    breed: String,
}

impl Dog {
    fn speak(&self) -> String {
        format!("{} (a {})", self.animal.speak(), self.breed)
    }
}

fn main() {
    println!("--- struct with impl ---");
    let rex = Animal::new("Rex");
    println!("{}", rex.speak());

    println!("\n--- explicit trait implementation, used generically ---");
    announce(&rex);

    println!("\n--- trait default method, not overridden ---");
    println!("{}", rex.greeting());

    println!("\n--- composition, not inheritance/embedding ---");
    let dog = Dog {
        animal: Animal::new("Fido"),
        breed: "Labrador".to_string(),
    };
    println!("{}", dog.speak());
    println!("accessing through the field explicitly: {}", dog.animal.name);
}
